src/java/org/apache/cassandra/db/guardrails/Threshold.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.guardrails;
+
+import java.util.function.Function;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+import org.apache.cassandra.service.ClientState;
+
+/**
+ * A guardrail based on numeric threshold(s).
+ *
+ * <p>A {@link Threshold} guardrail defines (up to) 2 threshold, one at which a warning is issued, and a higher one
+ * at which a failure is triggered. Only one of those threshold can be activated if desired.
+ *
+ * <p>This guardrail only handles guarding positive values.
+ */
+public class Threshold extends Guardrail
+{
+    /**
+     * A {@link Threshold} with both failure and warning thresholds disabled, so that cannot ever be triggered.
+     */
+    public static final Threshold NEVER_TRIGGERED = new Threshold("never_triggered", state -> -1L, state -> -1L, null);
+
+    private final Function<ClientState, Long> warnThreshold;
+    private final Function<ClientState, Long> failThreshold;
```

Comment:
this causes boxing, can we use `ToLongFunction` and call `applyAsLong`?

Reply:
Good idea, done.